src/utils.d

```diff
@@ -0,0 +1,125 @@
+// Compiler implementation of the D programming language
+// Copyright (c) 1999-2015 by Digital Mars
+// All Rights Reserved
+// written by Walter Bright
+// http://www.digitalmars.com
+// Distributed under the Boost Software License, Version 1.0.
+// http://www.boost.org/LICENSE_1_0.txt
+
+module ddmd.utils;
+
+import core.stdc.string;
+import ddmd.dmodule;
+import ddmd.dscope;
+import ddmd.errors;
+import ddmd.globals;
+import ddmd.id;
+import ddmd.identifier;
+import ddmd.parse;
+import ddmd.root.file;
+import ddmd.root.filename;
+import ddmd.root.outbuffer;
+import ddmd.tokens;
+
+extern (C++) __gshared Module entrypoint = null;
+extern (C++) __gshared Module rootHasMain = null;
+
+/************************************
+ * Generate C main() in response to seeing D main().
+ * This used to be in druntime, but contained a reference to _Dmain
+ * which didn't work when druntime was made into a dll and was linked
+ * to a program, such as a C++ program, that didn't have a _Dmain.
+ */
+extern (C++) void genCmain(Scope* sc)
+{
+    if (entrypoint)
+        return;
+    /* The D code to be generated is provided as D source code in the form of a string.
+     * Note that Solaris, for unknown reasons, requires both a main() and an _main()
+     */
+    static __gshared const(char)* cmaincode =
+    q{
+        extern(C)
+        {
+            int _d_run_main(int argc, char **argv, void* mainFunc);
+            int _Dmain(char[][] args);
+            int main(int argc, char **argv)
+            {
+                return _d_run_main(argc, argv, &_Dmain);
+            }
+            version (Solaris) int _main(int argc, char** argv) { return main(argc, argv); }
+        }
+    };
```

Comment:
> Any refactoring of the modules in the front end should act to reduce import cycles, not create new ones.

Yes, and I'm all for this too.  It is in my best interest that switching to the new D frontend won't add new source dependencies that I severed in the old C++ codebase.